Add flags to make listing easy to diff

lefessan opened the pull request:
follow-up to the discussion originally in https://github.com/OCamlPro/gnucobol/pull/75, implementing part of https://sourceforge.net/p/gnucobol/feature-requests/294/

New flags:
* -fno-ttimestamp: suppress time stamp
* -fttitle=<title>: display <title> instead of package name and version

GitMensch commented on tests/atlocal.in:
```diff
@@ -76,6 +76,7 @@ COBC="${COBC} -fdiagnostics-plain-output"
 COMPILE="${COBC} -x ${FLAGS}"
 COMPILE_ONLY="${COBC} -fsyntax-only ${FLAGS} -Wno-unsupported"
 COMPILE_MODULE="${COBC} -m ${FLAGS}"
+COMPILE_LISTING="${COMPILE_ONLY} -ftversion=V.R.P -fno-ttimestamp -tlines=0"
```
please go with
`COMPILE_LISTING="${COMPILE_ONLY} -fttitle=\"GnuCOBOL V.R.P\" -fno-ttimestamp -tlines=0"`
there should be no issues with escaping and while I see the developer ease of `-ftversion` that should not get into the release

lefessan replied:
I dropped the `-ftversion`. However, it is not possible to pass an argument with a space through a command stored in an environment variable in a portable way. So, I modified the behavior of `-fttitle` so that underscores in `-fttitle=<...>` are replaced by spaces.

GitMensch commented on tests/testsuite.src/listings.at:
```diff
@@ -97,9 +97,7 @@ AT_DATA([prog.cob], [
-AT_CHECK([$COMPILE_ONLY -t prog.lst -tlines=0 prog.cob], [0], [], [])
-
-AT_DATA([prog1.lst],
+AT_CHECK([$COMPILE_LISTING -t- prog.cob], [0],
```
I like that very much and given that cobc normally outputs to stderr this is also nicely to handle

GitMensch commented on cobc/ChangeLog:
```diff
@@ -1,4 +1,11 @@
+2023-07-02  Fabrice Le Fessant <fabrice.le_fessant@ocamlpro.com>
+
+	* flag.def/cobc.c: new flags -fno-ttimestamp to suppress timestamp
+	in listing headers, and -fttitle=<TITLE>, so that listings can be
```
please 2-space indent after the asterisk as in the other entries

lefessan replied:
done

GitMensch commented on cobc/cobc.c:
```diff
@@ -90,6 +90,24 @@ enum compile_level {
 	CB_LEVEL_EXECUTABLE	= 7
 };
 
+#define	CB_FLAG_GETOPT_STACK_SIZE       1
```
Maybe those should be in flag.def, directly before the matching CB_FLAG or at its header?

lefessan replied:
`flag.def` is included several times into `cobc.c`, so I was not confortable with these macros being included several times, maybe generating warnings on some systems.

GitMensch replied:
That multi-inclusion is totally right, so cobc.c is the right place.

GitMensch commented on cobc/cobc.c:
```diff
@@ -3720,16 +3748,17 @@ process_command_line (const int argc, char **argv)
 			break;
 
-		case 6:
-			/* -fdefaultbyte=<xx> : Default initialization byte */
```
if not used anymore and therefore deleted, please note that in the Changelog

lefessan replied:
done

GitMensch reviewed (changes requested):
This could go live tomorrow - but `-ftversion` is really out of the question, so please adjust.

And just as a note: we don't need to add the other two to NEWS; but as this is a global change to testsuite.src/listing.at this is one of the cases where a Changelog entry for the testsuite is good.

GitMensch commented:
Just stumbled over that: one of the listing tests has 2x identical prog.cob with `PROCEDURE DIVISION USING PAR-IN RETURNING PAR-OUT.` - if possible please drop the second AT_DATA when adjusting listing.at

lefessan committed:
Add flags to make listing easy to diff

New flags:
    -fno-ttimestamp: suppress time stamp
    -fttitle=<title>: display <title> instead of package name and version
Partially implements feature request #294

lefessan commented on cobc/cobc.c:
```diff
@@ -2250,8 +2274,12 @@ set_listing_date (void)
+	if (cb_listing_with_timestamp){
+		strftime (cb_listing_date, (size_t)CB_LISTING_DATE_MAX,
+			  LISTING_TIMESTAMP_FORMAT, &current_compile_tm);
+	} else {
```
fixed

GitMensch replied:
I suggest to leave this code unchanged (= original) and instead check `cb_listing_with_timestamp` in `print_program_header()`.
The current implementation would generate a trailing space for `--tlines=0` which may be a reason for the testsuite failures you currently see.

lefessan commented:
> Just stumbled over that: one of the listing tests has 2x identical prog.cob with `PROCEDURE DIVISION USING PAR-IN RETURNING PAR-OUT.` - if possible please drop the second AT_DATA when adjusting listing.at

I think I found it and removed it.

lefessan commented:
Argh, tests run fine with `autofonce` but not with `autotest`, I need to investigate the different behavior...

GitMensch commented on cobc/cobc.c:
```diff
@@ -3720,16 +3745,20 @@ process_command_line (const int argc, char **argv)
+			/* -fttitle=<title> : Title for listing */
+			{
+				size_t i, len;
+				cb_listing_with_title = cobc_main_strdup (cob_optarg);
```
misses the check if it is already set, then free that we have for `-t` as the memory is not "leaked" we may drop this later but should be consistent within the same function about that

lefessan replied:
done

GitMensch committed:
Update flag.def

minor spacing issue and dropped info from help that never was true

GitMensch commented on tests/atlocal.in:
```diff
@@ -76,6 +76,9 @@ COBC="${COBC} -fdiagnostics-plain-output"
 COMPILE="${COBC} -x ${FLAGS}"
 COMPILE_ONLY="${COBC} -fsyntax-only ${FLAGS} -Wno-unsupported"
 COMPILE_MODULE="${COBC} -m ${FLAGS}"
+LISTING_FLAGS="-fttitle=GnuCOBOL_V.R.P -fno-ttimestamp"
```
Shouldn't `-t -` be in there, too? For `AT_CHECK we still can override this by `-T -` for the tests that check the wide listing.
I'll leave that decision to you, just wanted to bring it up.

lefessan replied:
There are some places where we still need to output to a file. I will keep it outside for now.

GitMensch reviewed (changes requested):
only minor changes needed (place where `cb_listing_with_timestamp` is read), which may also fix the tests - as soon as this is done and the test run - feel free to commit to svn, so we can go on with the REPLACE issue

one thing that is not mandatory but may be done in this PR: move `--tlines` to `-ftlines` (drop from help.c, keep setting in cobc.c as compatibility for now, only the definition moves to flag.def)

Note: I took the freedom to adjust the help output in flags.def.

GitMensch commented:
> Argh, tests run fine with `autofonce` but not with `autotest`, I need to investigate the different behavior...

Just inspected the testsuite.log: it is exactly as I've thought:


```diff
-GnuCOBOL V.R.P               prog.cob
+GnuCOBOL V.R.P               prog.cob                   
```

To fix that change the place where `cb_listing_with_timestamp` is checked (should be in `print_program_header()`) and the tests will pass. Note that this directly allows to give either the title or the filename more space, too.

lefessan commented:
I will commit this version to SVN as soon as the CI passes, to be able to switch to the other PR.
There are still tests with spaces at the end of headers, that I fixed using a quadrigraph (`@&t@`) to prevent autoconf from removing them. I preferred not to change the printing because the spaces are generated by `%-26.26s` in `printf` and I was not sure of the intention.

lefessan commented:
( I just pushed a commit in `autofonce` to correctly handle spaces, in case you use it)

GitMensch commented:
> I preferred not to change the printing because the spaces are generated by `%-26.26s` in `printf` and I was not sure of the intention.

The point was that there is a trailing timestamp, right adjusted.
If `-fno-ttimestamp` is used we currently printf the low-value; but when this is changed to not adjustt the timestamp var but instead adjust the printing and right-justify the sourcename, then we should not have any trailing spaces left.

GitMensch commented:
LGTM (we can right-adjust the filename for `-fno-ttimestamp` later)

lefessan commented:
Merged with whitespaces removed at the end of line
